lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+                # check the filesystem for it.
+                pkg.stage.created = False
+
+            # Perform basic task cleanup for the installed spec to
+            # include downgrading the write to a read lock
+            self._cleanup_task(pkg, True)
+
+        # Cleanup, which includes releasing all of the read locks
+        self._cleanup_all_tasks()
+
+        # Ensure we properly report if the original/explicit pkg is failed
+        if self.pkg.unique_id in self.failed:
+            msg = ('Installation of {0} failed.  Review log for details'
+                   .format(self.pkg.unique_id))
+            tty.error(msg)
+            raise InstallError(msg)
+
+    install.__doc__ += install_args_docstring
+
+    # Helper method to "smooth" the transition from the
+    # spack.package.PackageBase class
+    @property
+    def spec(self):
+        """The specification associated with the package."""
+        return self.pkg.spec
+
+
+class BuildTask(object):
+    """Class for representing the build task for a package."""
+
+    def __init__(self, pkg, compiler, start, attempts, status, installed):
+        """
+        Instantiate a build task for a package.
+
+        Args:
+            pkg (Package): the package to be installed and built
+            compiler (bool): ``True`` if the task is for a bootstrap compiler,
+                otherwise, ``False``
+            start (int): the initial start time for the package, in seconds
+            attempts (int): the number of attempts to install the package
+            status (int): the installation status
+            installed (list of str): the identifiers of packages that have
+                been installed so far
+        """
+
+        # Ensure dealing with a package that has a concretized spec
+        if isinstance(pkg, spack.package.PackageBase):
+            self.pkg = pkg
+            if not self.pkg.spec.concrete:
+                raise ValueError("{0} must have a concrete spec"
+                                 .format(self.pkg.unique_id))
+        else:
+            raise ValueError("{0} must be a package".format(str(pkg)))
+
+        # Initialize the status to an active state.  The status is used to
+        # ensure priority queue invariants when tasks are "removed" from the
+        # queue.
+        if status != STATUS_REMOVED:
+            self.status = status
+        else:
+            msg = "Cannot create a build task for {0} with status '{1}'"
+            raise InstallError(msg)
+
+        # Package is associated with a bootstrap compiler
+        self.compiler = compiler
+
+        # The initial start time for processing the spec.
+        # TODO: Should this be tied to the status?
+        self.start = start
+
+        # Number of times the task has been queued
+        self.attempts = attempts + 1
+
+        # Set of dependents
+        self.dependents = set(d.package.unique_id for d
+                              in self.spec.dependents())
```

Comment:
Is this safe in view of bug #11983?

Reply:
That bug _may_  explain why I had to switch from including the DAG hash in the `unique_id` in commit `50216fd`.